dali/kernels/common/join/tensor_join_cpu_test.cc

```diff
@@ -0,0 +1,217 @@
+                                     {4, 5, 3, 8}};
+  TensorShape<> shout = {4, 5, 10, 8};
+  EXPECT_EQ(impl::DetermineShape<false>(make_span(shin), 2), shout);
+}
+
+
+TEST(TensorJoinCpuTest, ConcatenateTensorsTest) {
+  using namespace std;  // NOLINT
+
+  vector<vector<int>> arr = {{6, 8, 5, 1, 3, 5, 1, 6, 8, 3, 7, 5},
+                             {4, 5, 1, 8, 4, 4, 1, 4, 1, 7, 6, 6}};
+  vector<TensorShape<>> sh = {{3, 4},
+                              {3, 4}};
+  vector<TensorView<StorageCPU, const int>> in;
+  for (size_t i = 0; i < arr.size(); i++) {
+    in.emplace_back(arr[i].data(), sh[i]);
+  }
+
+  // Output shape for this buffer: {2, 3, 4} (STACK) or {6, 4} (CONCAT)
+  vector<int> arr0 = {6, 8, 5, 1, 3, 5, 1, 6, 8, 3, 7, 5, 4, 5, 1, 8, 4, 4, 1, 4, 1, 7, 6, 6};
+
+  // Output shape for this buffer: {3, 2, 4} (STACK) or {3, 8} (CONCAT)
+  vector<int> arr1 = {6, 8, 5, 1, 4, 5, 1, 8, 3, 5, 1, 6, 4, 4, 1, 4, 8, 3, 7, 5, 1, 7, 6, 6};
+
+  // Output shape for this buffer: {4, 3, 2} (STACK), CONCAT unavailable
+  vector<int> arr2 = {6, 4, 8, 5, 5, 1, 1, 8, 3, 4, 5, 4, 1, 1, 6, 4, 8, 1, 3, 7, 7, 6, 5, 6};
+
+  vector<vector<int>> ref_arr;
+  ref_arr.emplace_back(arr0);
+  ref_arr.emplace_back(arr1);
+  ref_arr.emplace_back(arr2);
+
+  for (size_t ax = 0; ax < ref_arr.size(); ax++) {
+    auto outsh = impl::DetermineShape<true>(make_span(sh), ax);
+    vector<int> outbuf(volume(outsh));
+    ASSERT_EQ(volume(outsh), ref_arr[ax].size());
+    impl::ConcatenateTensors({outbuf.data(), outsh}, make_cspan(in), ax);
+    EXPECT_EQ(outbuf, ref_arr[ax]);
+  }
+
+  for (size_t ax = 0; ax < ref_arr.size() - 1; ax++) {
+    auto outsh = impl::DetermineShape<false>(make_span(sh), ax);
+    vector<int> outbuf(volume(outsh));
+    ASSERT_EQ(volume(outsh), ref_arr[ax].size());
+    impl::ConcatenateTensors({outbuf.data(), outsh}, make_cspan(in), ax);
+    EXPECT_EQ(outbuf, ref_arr[ax]);
+  }
+}
+
+
+template<template<typename, int = -1> class Kernel, typename T>
+void
+KernelRunAndVerify(TensorView<StorageCPU, T> ref, span<const TensorView<StorageCPU, const T>> in,
+                   int axis) {
+  Kernel<T> kernel;
+  KernelContext c;
+  auto kr = kernel.Setup(c, in, axis);
+  auto outsh = kr.output_shapes[0].tensor_shape(0);
+  std::vector<T> outbuf(volume(outsh));
+  TensorView<StorageCPU, T> out{outbuf.data(), outsh};
+  kernel.Run(c, out, in);
+  ASSERT_EQ(out.shape, ref.shape);
+  if (std::is_integral<T>::value) {
+    for (int i = 0; i < volume(ref.shape); i++) {
+      EXPECT_EQ(out.data[i], ref.data[i]);
+    }
+  } else {
+    for (int i = 0; i < volume(ref.shape); i++) {
+      EXPECT_FLOAT_EQ(out.data[i], ref.data[i]);
+    }
+  }
+}
+
+
+TEST(TensorJoinCpuTest, StackKernelTest) {
+  using namespace std;  // NOLINT
+  vector<vector<int>> arr = {{100, 101, 102, 110, 111, 112, 120, 121, 122, 130, 131, 132},
+                             {200, 201, 202, 210, 211, 212, 220, 221, 222, 230, 231, 232}};
```

Comment:
Sorry for being nitpicky, but... can't you format these into 3x4 arrays? It would be so much easier to read.